### PR TITLE
    pull: Add support for basic auth

### PR DESCRIPTION
**Pull request by cgwalters**
This has been requested a few times for people delivering
OSTree content and wanting to do access control.

### CODE REVIEW

**Comment by kleest** [comment]
Any update on this?

**Comment by Mydayyy** [comment]
Hello kleest,

we had the same problem. As the author of this commit fell asleep last september we searched for a workaround. Luckily, we just found one. 

You can just place the basic auth on the server inside the url for ostree 

Regards
Mydayyy ( Someone who actually finishes pull requests )

File: src/libostree/ostree-fetcher-curl.c [hunk 1]
```diff
@@ -771,6 +771,9 @@ initiate_next_curl_request (FetcherRequest *req,
   if ((self->config_flags & OSTREE_FETCHER_FLAGS_TRANSFER_GZIP) > 0)
     curl_easy_setopt (req->easy, CURLOPT_ACCEPT_ENCODING, "");
 
+  /* If we have e.g. basic auth in the URL string, let's honor that */
+  curl_easy_setopt (req->easy, CURLOPT_HTTPAUTH, CURLAUTH_ANY);
```
Comment (dnohales) [inline src/libostree/ostree-fetcher-curl.c]: Take note that when the URL is set to the curl object, at some point we call `soup_uri_to_string`, which on the [documentation](https://developer.gnome.org/libsoup/stable/SoupURI.html#soup-uri-to-string) specifically says:

> Note that the output will never contain a password, even if `uri` does.

I'd suggest to set the username and password manually like this:

```suggestion
  const char *username = soup_uri_get_user (req->mirrorlist->pdata[req->idx]);
  curl_easy_setopt (req->easy, CURLOPT_USERNAME, username);

  const char *password = soup_uri_get_password (req->mirrorlist->pdata[req->idx]);
  curl_easy_setopt (req->easy, CURLOPT_PASSWORD, password);
```

With this change it works for me for a protected repo.

**Comment by dnohales** [comment]
Any updates?

File: src/libostree/ostree-fetcher-curl.c (hunk 1 above)
Comment (palvarez89) [reply src/libostree/ostree-fetcher-curl.c]: This worked for me too, thanks!!

**Comment by jlebon** [comment]
@cgwalters This code looks pretty good to me! WDYT about taking the suggested change and just getting this in?

File: src/libostree/ostree-fetcher-curl.c (hunk 1 above)
Comment (cgwalters) [reply src/libostree/ostree-fetcher-curl.c]: Thanks a lot for taking the time to track that down!  I ended up using that suggestion but factored out a variable for the URI that we use once.

**Comment by cgwalters** [comment]
OK rebased :surfing_man:  and updated with (tweaked) suggestion.

**Pull request merged**